File: src/main/java/algorithms/java/easy/ReverseInteger.java
package algorithms.java.easy;

/**
 * #7：ReverseInteger
 * https://leetcode-cn.com/problems/reverse-integer/
 * 时间复杂度：O(logn)
 * 思路：
 * 1 通过%10将整数从尾数弹出放入r中
 * 2 x弹完=0时，返回r（判断溢出）
 */
public class ReverseInteger {

    public static int reverse(int x) {
        long r = 0;
        while(x != 0) {
            int pop = x % 10;       //从x中弹出最后一位
            x = x / 10;             //改变x值
            //防止 r 溢出
//            if(r > Integer.MAX_VALUE / 10 || (r == Integer.MAX_VALUE / 10 && pop > 7)) return 0;
//            if(r < Integer.MIN_VALUE / 10 || (r == Integer.MIN_VALUE / 10 && pop < -8)) return 0;
            r = r * 10 + pop;       //r将x弹出的值追加在尾部
        }
        return r > Integer.MAX_VALUE || r < Integer.MIN_VALUE ? 0 : (int) r;
    }

    public static void main(String[] args) {
        int x = 12345678;
        System.out.println(ReverseInteger.reverse(x));
    }
}

File: ruby/reverse_integer.rb
=begin
 * LC#7：ReverseInteger 整数反转
 * https://leetcode-cn.com/problems/reverse-integer/
 * 思路1（runtime beats 最优解）：通过%10将整数从尾数弹出放入r中，x弹完=0时，返回r（判断溢出）
=end

def reverse_integer (target_num)
  # ruby 负数求余有些问题，转成正数后再处理
  is_negative = true if target_num < 0
  target_num = target_num.abs
  r = 0
  while target_num > 0 do
    r = r * 10 + (target_num % 10)
    target_num = target_num / 10
    r = 0 if r > (2**31) || r < (-2**31)
  end
  is_negative ? r * -1 : r
end

r = reverse_integer(1534236469)
p "r >", r